airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -190,32 +199,120 @@ PARTITION BY (DATE_TRUNC(_airbyte_extracted_at, DAY))
-     */
-    return "";
+    LOGGER.info("Alter Table Report {} {} {} {}; Clustering {}; Partitioning {}",
+            alterTableReport.columnsToAdd(),
+            alterTableReport.columnsToRemove(),
+            alterTableReport.columnsToChangeType(),
+            alterTableReport.isDestinationV2Format(),
+            tableClusteringMatches,
+            tablePartitioningMatches);
+
+    return alterTableReport.isNoOp() && tableClusteringMatches && tablePartitioningMatches;
+  }
+
+  @VisibleForTesting
+  public boolean clusteringMatches(StreamConfig stream, StandardTableDefinition existingTable) {
+    return existingTable.getClustering() == null ? false : containsAllIgnoreCase(
+            existingTable.getClustering().getFields().stream().collect(Collectors.toSet()),
+            clusteringColumns(stream));
+  }
+
+  @VisibleForTesting
+  public boolean partitioningMatches(StandardTableDefinition existingTable) {
+    return existingTable.getTimePartitioning() == null ? false : existingTable.getTimePartitioning()
+            .getField()
+            .equalsIgnoreCase("_airbyte_extracted_at") &&
+            TimePartitioning.Type.DAY.equals(existingTable.getTimePartitioning().getType());
+  }
+
+  @Override
+  public AlterTableReport buildAlterTableReport(final StreamConfig stream, final TableDefinition existingTable) {
+    final Map<String, StandardSQLTypeName> streamSchema = stream.columns().entrySet().stream()
+            .collect(Collectors.toMap(
+                    entry -> entry.getKey().name(),
+                    entry -> toDialectType(entry.getValue())
+            ));
+
+    final Map<String, StandardSQLTypeName> existingSchema = existingTable.getSchema().getFields().stream()
+            .collect(Collectors.toMap(
+                    field -> field.getName(),
+                    field -> field.getType().getStandardType()
+            ));
+
+    // Columns in the StreamConfig that don't exist in the TableDefinition
+    final Set<String> columnsToAdd = streamSchema.keySet().stream()
+            .filter(name -> !containsIgnoreCase(existingSchema.keySet(), name))
+            .collect(Collectors.toSet());
+
+    // Columns in the current schema that are no longer in the StreamConfig
+    final Set<String> columnsToRemove = existingSchema.keySet().stream()
+            .filter(name -> !containsIgnoreCase(streamSchema.keySet(), name) && !containsIgnoreCase(FINAL_TABLE_AIRBYTE_COLUMNS, name))
+            .collect(Collectors.toSet());
+
+    // Columns that are typed differently than the StreamConfig
+    final Set<String> columnsToChangeType = streamSchema.keySet().stream()
+            // If it's not in the existing schema, it should already be in the columnsToAdd Set
+            .filter(name -> {
+              // Big Query Columns are case-insensitive
+              final String matchedKey = matchingKey(existingSchema.keySet(), name).orElse(name);
```

Comment:
I found this lambda kind of hard to follow, maybe better to be explicit:
```java
Optional<String> matchedKey = matchingKey(existingSchema.keySet(), name);
if (!matchedKey.isPresent()) { return false; }
return !existingSchema.get(matchedKey.get()).equals(streamSchema.get(name));
```

Reply:
How does this look
```java
    // Columns that are typed differently than the StreamConfig
    final Set<String> columnsToChangeType = streamSchema.keySet().stream()
            // If it's not in the existing schema, it should already be in the columnsToAdd Set
            .filter(name -> {
              // Big Query Columns are case-insensitive, first find the correctly cased key if it exists
              return matchingKey(existingSchema.keySet(), name)
                      // if it does exist, only include it in this set if the type (the value in each respective map)
                      // is different between the stream and existing schemas
                      .map(key -> !existingSchema.getOrDefault(key, streamSchema.get(name)).equals(streamSchema.get(name)))
                      // if there is no matching key, then don't include it because it is probably already in columnsToAdd
                      .orElse(false);
            })
            .collect(Collectors.toSet());
```

Reply:
is the getOrDefault still needed? since we're only entering that map lambda if we found column in existingSchema

but yeah this makes a lot more sense!

Reply:
oh yeah good call!